03_oop_part_2/src/item.ts

```diff
@@ -0,0 +1,33 @@
+import { Page } from './page';
+import { Pages } from './pages';
+
+
+const PagesIterable = () => class implements IterableIterator<Item> {
+    constructor(protected pages: Pages, protected openPage?: Page) {}
+
+    public next(): IteratorResult<Item> {
+        let nxt = this.pages.next();
+        if (!nxt.done) {
+            this.openPage = nxt.value;
+            return {done: false, value: this};
+        }
+        this.openPage = undefined;
+        return {done: true, value: null};
+    }
+
+    [Symbol.iterator](): IterableIterator<Item> {
+        return this;
+    }
+
+    protected get pageSuffix(): String {
+        if (typeof(this.openPage) === 'object') {
+            return  `, ${this.openPage}`;
+        }
+        return '';
+    }
+}
+
+export abstract class Item extends PagesIterable() {
```

Comment:
I suggest to use the mixin in a more natural form:
```ts
export const PagesIterableItem = PagesIterable(Item)
```
Thus we can apply some constraints on a class passed to `PagesIterable`.

`Book` and other classes will extend `PagesIterableItem` class (and transitively `Item` class as well).